boot/systems_test.go

```diff
@@ -214,15 +215,155 @@ func (s *systemsSuite) TestSetTryRecoverySystemEncrypted(c *C) {
 	}), Equals, true)
 }
 
+func (s *systemsSuite) TestSetTryRecoverySystemRemodelEncrypted(c *C) {
+	mockAssetsCache(c, s.rootdir, "trusted", []string{
+		"asset-asset-hash-1",
+	})
+
+	mtbl := s.mockTrustedBootloaderWithAssetAndChains(c, s.runKernelBf, s.recoveryKernelBf)
+	bootloader.Force(mtbl)
+	defer bootloader.Force(nil)
+
+	// system is encrypted
+	s.stampSealedKeys(c, s.rootdir)
+
+	model := s.uc20dev.Model()
+	newModel := boottest.MakeMockUC20Model(map[string]interface{}{
+		"model": "my-new-model",
+	})
+
+	modeenv := &boot.Modeenv{
+		Mode: "run",
+		// keep this comment to make old gofmt happy
+		CurrentRecoverySystems: []string{"20200825"},
+		GoodRecoverySystems:    []string{"20200825"},
+		CurrentTrustedRecoveryBootAssets: boot.BootAssetsMap{
+			"asset": []string{"asset-hash-1"},
+		},
+		CurrentTrustedBootAssets: boot.BootAssetsMap{
+			"asset": []string{"asset-hash-1"},
+		},
+		CurrentKernels: []string{"pc-kernel_500.snap"},
+
+		Model:          model.Model(),
+		BrandID:        model.BrandID(),
+		Grade:          string(model.Grade()),
+		ModelSignKeyID: model.SignKeyID(),
+	}
+	c.Assert(modeenv.WriteTo(""), IsNil)
+
+	var readSeedSeenLabels []string
+	restore := boot.MockSeedReadSystemEssential(func(seedDir, label string, essentialTypes []snap.Type, tm timings.Measurer) (*asserts.Model, []*seed.Snap, error) {
+		// the mock bootloader can only mock a single recovery boot
+		// chain, so pretend both seeds use the same kernel, but keep track of the labels
+		readSeedSeenLabels = append(readSeedSeenLabels, label)
+		systemModel := model
+		if label == "1234" {
+			systemModel = newModel
+		}
+		return systemModel, []*seed.Snap{s.seedKernelSnap, s.seedGadgetSnap}, nil
+	})
+	defer restore()
+
+	resealCalls := 0
+	restore = boot.MockSecbootResealKeys(func(params *secboot.ResealKeysParams) error {
+		resealCalls++
+		// bootloader variables have already been modified
+		c.Check(mtbl.SetBootVarsCalls, Equals, 1)
+		c.Assert(params, NotNil)
+		switch resealCalls {
+		case 1:
+			c.Assert(params.ModelParams, HasLen, 2)
+			c.Check(params.KeyFiles, DeepEquals, []string{
+				filepath.Join(boot.InitramfsBootEncryptionKeyDir, "ubuntu-data.sealed-key"),
+			})
+			c.Assert(params.ModelParams[0].KernelCmdlines, DeepEquals, []string{
+				"snapd_recovery_mode=recover snapd_recovery_system=20200825 static cmdline",
+				"snapd_recovery_mode=run static cmdline",
+			})
+			c.Assert(params.ModelParams[1].KernelCmdlines, DeepEquals, []string{
+				"snapd_recovery_mode=recover snapd_recovery_system=1234 static cmdline",
+				"snapd_recovery_mode=run static cmdline",
```

Comment:
strictly speaking we don't need to seal for run mode this system, right? not saying that it's easy to stop or a is a bit issue, but wondering

Reply:
Yes. I think we can optimize this a little bit once things land